// @trpc/client@11.0.0-alpha-next-2023-12-18-04-17-50.120/dist/links/httpLink.mjs
import { observable } from '@trpc/server/observable';
import { t as transformResult } from '../transformResult-c5af197e.mjs';
import { T as TRPCClientError } from '../TRPCClientError-5f32366f.mjs';
import { r as resolveHTTPLinkOptions, j as jsonHttpRequester } from '../httpUtils-5c68b5c4.mjs';

function httpLinkFactory(factoryOpts) {
    return (opts)=>{
        const resolvedOpts = resolveHTTPLinkOptions(opts);
        return (runtime)=>({ op  })=>observable((observer)=>{
                    const { path , input , type  } = op;
                    const { promise , cancel  } = factoryOpts.requester({
                        ...resolvedOpts,
                        runtime,
                        type,
                        path,
                        input,
                        headers () {
                            if (!opts.headers) {
                                return {};
                            }
                            if (typeof opts.headers === 'function') {
                                return opts.headers({
                                    op
                                });
                            }
                            return opts.headers;
                        }
                    });
                    let meta = undefined;
                    promise.then((res)=>{
                        meta = res.meta;
                        const transformed = transformResult(res.json, runtime);
                        if (!transformed.ok) {
                            observer.error(TRPCClientError.from(transformed.error, {
                                meta
                            }));
                            return;
                        }
                        observer.next({
                            context: res.meta,
                            result: transformed.result
                        });
                        observer.complete();
                    }).catch((cause)=>{
                        observer.error(TRPCClientError.from(cause, {
                            meta
                        }));
                    });
                    return ()=>{
                        cancel();
                    };
                });
    };
}
/**
 * @see https://trpc.io/docs/client/links/httpLink
 */ const httpLink = httpLinkFactory({
    requester: jsonHttpRequester
});

export { httpLink, httpLinkFactory };
